### ISSUE TITLE
In Flutter child navigator, Android OS back button will pop the whole child stack

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#55570]
Internal: b/154569321

This seems to be by design. The only place in Flutter framework that overrides `didPopRoute()` is the WidgetsAppState. I would have thought the `Navigator` widget would also have a WidgetsBinding observer registered and would respond to OS back button. That's not the case.

Use the sample app `flutter create --sample=widgets.Navigator.1 mysample` and observe that when we are at "Choose Credentials Page" (page 2 of the inner navigator), clicking back button pops us all the way back to homepage instead of going back to  page 1 of the inner navigator.

There's a workaround for this but it is not trivial. One would have to convert the widget creating the child navigator to a stateful widget and pass a unique key to the navigator so its state can be locate in the widget tree when didPopRoute is called. We should likely handle this automatically.

**Comment by goderbauer** [comment]
The Navigator 2.0 design introduces a BackButtonDispatcher that allows you to choose which Navigator you want to pop when the system back button is pressed.

/cc @chunhtai, who is implementing Nav 2.0

**Comment by chunhtai** [comment]
While this problem can be solved in navigator 2.0, there is the [willpopscope](https://api.flutter.dev/flutter/widgets/WillPopScope-class.html) that you can veto the pop. You can wrap the child navigator with this widget and handles the pop in there. This might make this slightly easier.

**Comment by chunhtai** [comment]
Is this still an issue if uses the willpopscope?

**Comment by chunhtai** [comment]
The new router widget will be added in this pr https://github.com/flutter/flutter/pull/60299

**Comment by chunhtai** [comment]
If you are using a router widget, you will be able to achieve dispatch the backbutton with the following code

```dart
class MyChildRouter extends StatelessWidget {

  @override
   Widget build(BuildContext context) {
       // Retrieve the back button dispatcher from the parent.
       final BackButtonDispatcher parentDispatcher = Router.of(context).backButtonDispatcher;
       
       final BackButtonDispatcher childDispatcher = parentDispatcher. createChildBackButtonDispatcher();
       // tells the parent dispatcher to delegate the back button intend to this child.
       innerDispatcher.takePriority();
       
        return Router(
           ...
           backbuttonDispatcher: childDispatcher,
       );
   }
}
```

**Comment by pedromassangocode** [comment]
Does this has been solved with Navigator 2.0? So that this can be closed.

**Comment by nguyenxndaidev** [comment]
@chunhtai 

> If you are using a router widget, you will be able to dispatch the back button intend with the following code.

But using router widget, you will need to create `RouterDelegate` and introduce unnecessary complexity to the app, which doesn't target the web or deeplink.

It would be nicer if `Navigator` also support `BackButtonDispatcher` or similar way.
